File: Mediations and Networks/KEGG_Annotation.py
import os
import csv
from collections import Counter
from concurrent.futures import ThreadPoolExecutor
from kegg_pull.rest import KEGGrest

# List of input files
file_paths = [
    "drug_non_drug_base_KEGG_KO_stats.csv",
    "drug_non_drug_0_base_KEGG_KO_stats.csv",
    "drug_non_drug_base_KEGG_pathway_stats.csv",
    "drug_non_drug_0_base_KEGG_pathway_stats.csv",
    "drug_non_drug_base_KEGG_module_stats.csv",
    "drug_non_drug_0_base_KEGG_module_stats.csv"
]

# Collect all DisplayNames
all_display_names = []
for file in file_paths:
    if os.path.exists(file):
        with open(file, newline='') as f:
            reader = csv.DictReader(f)
            if 'DisplayName' in reader.fieldnames:
                all_display_names.extend(row['DisplayName'] for row in reader)
            else:
                print(f"Warning: 'DisplayName' column not found in {file}")
    else:
        print(f"Warning: File not found: {file}")

# Find duplicates
duplicates = [item for item, count in Counter(all_display_names).items() if count > 1]

# KEGG functions
kegg = KEGGrest()

def get_kegg_name(entry_id, prefix):
    try:
        response = kegg.get(entry_ids=f"{prefix}:{entry_id}")
        if response.status.successful:
            for line in response.text_body.splitlines():
                if line.startswith("NAME"):
                    return line.replace("NAME", "").strip()
        return "not found"
    except Exception:
        return "not found"

def get_kegg_links(target, source_id, prefix):
    try:
        response = kegg.link(target_database=target, source_database=f"{prefix}:{source_id}")
        if response.status.successful:
            ids = [line.split("\t")[1].split(":")[1] for line in response.text_body.strip().splitlines()]
            return sorted(set(ids))
        return []
    except Exception:
        return []

def annotate_entry(entry_id):
    if entry_id.startswith("K"):
        entry_type = "KO"
        name = get_kegg_name(entry_id, "ko")
        modules = get_kegg_links("module", entry_id, "ko")
        pathways = get_kegg_links("pathway", entry_id, "ko")
        module_names = [get_kegg_name(mid, "md") for mid in modules]
        pathway_names = [get_kegg_name(pid, "path") for pid in pathways]
    elif entry_id.startswith("M"):
        entry_type = "Module"
        name = get_kegg_name(entry_id, "md")
        modules = [entry_id]
        module_names = [name]
        pathways = get_kegg_links("pathway", entry_id, "md")
        pathway_names = [get_kegg_name(pid, "path") for pid in pathways]
    elif entry_id.startswith("map") or entry_id.startswith("path"):
        entry_type = "Pathway"
        name = get_kegg_name(entry_id, "path")
        modules = []
        module_names = []
        pathways = [entry_id]
        pathway_names = [name]
    else:
        entry_type = "Unknown"
        name = "unknown"
        modules = []
        module_names = []
        pathways = []
        pathway_names = []

    return {
        "Input_ID": entry_id,
        "Type": entry_type,
        "Name": name,
        "Modules": "; ".join(modules),
        "Module_Names": "; ".join(module_names),
        "Pathways": "; ".join(pathways),
        "Pathway_Names": "; ".join(pathway_names)
    }

# Parallel annotation
with ThreadPoolExecutor() as executor:
    results = list(executor.map(annotate_entry, duplicates))

# Write output
with open("kegg_hierarchy_output.csv", "w", newline='') as f:
    fieldnames = ["Input_ID", "Type", "Name", "Modules", "Module_Names", "Pathways", "Pathway_Names"]
    writer = csv.DictWriter(f, fieldnames=fieldnames)
    writer.writeheader()
    writer.writerows(results)

print("KEGG annotation completed and saved to kegg_hierarchy_output.csv.")

File: Mediations and Networks/KEGG Annotation.R
# Load required packages
if (!requireNamespace("KEGGREST", quietly = TRUE)) {
  install.packages("BiocManager")
  BiocManager::install("KEGGREST")
}
library(KEGGREST)
library(readr)

# List of file paths
file_paths <- c(
  "drug_non_drug_base_KEGG_KO_stats.csv",
  "drug_non_drug_0_base_KEGG_KO_stats.csv",
  "drug_non_drug_base_KEGG_pathway_stats.csv",
  "drug_non_drug_0_base_KEGG_pathway_stats.csv",
  "drug_non_drug_base_KEGG_module_stats.csv",
  "drug_non_drug_0_base_KEGG_module_stats.csv"
)

# Initialize a list to store DisplayName values
all_display_names <- list()

# Read each file, filter Rank == 1, and extract DisplayName
for (file in file_paths) {
  if (file.exists(file)) {
    df <- read.csv(file, stringsAsFactors = FALSE)
    
    # Filter for Rank == 1 if column exists
    if ("Rank" %in% colnames(df)) {
      df <- df[df$Rank == 1, ]
    }
    
    # Extract DisplayName if present
    if ("DisplayName" %in% colnames(df)) {
      all_display_names[[file]] <- df$DisplayName
    } else {
      warning(paste("DisplayName column not found in", file))
    }
  } else {
    warning(paste("File not found:", file))
  }
}

# Combine and find duplicates
combined_display_names <- unlist(all_display_names)
duplicate_names <- unique(combined_display_names[duplicated(combined_display_names)])
kegg <- data.frame(duplicate_names, stringsAsFactors = FALSE)

# Define row indices for each type (adjust as needed)
ko_rows <- c(1:439)
module_rows <- c(461:513)
pathway_rows <- c(440:460)

# Helper: Get KEGG name
get_kegg_name <- function(id) {
  tryCatch({
    entry <- keggGet(id)[[1]]
    if (!is.null(entry$NAME)) entry$NAME else "unknown"
  }, error = function(e) "not found")
}

# Helper: Get linked IDs
get_kegg_links <- function(target, source_id) {
  tryCatch({
    links <- keggLink(target, source_id)
    unique(sub(paste0(target, ":"), "", links))
  }, error = function(e) NA)
}

# Process each ID
results <- lapply(seq_len(nrow(kegg)), function(i) {
  id <- kegg$duplicate_names[i]
  if (i %in% ko_rows) {
    ko_id <- paste0("ko:", id)
    list(
      Input_ID = id,
      Type = "KO",
      Name = get_kegg_name(ko_id),
      Modules = paste(get_kegg_links("module", ko_id), collapse = "; "),
      Pathways = paste(get_kegg_links("pathway", ko_id), collapse = "; ")
    )
  } else if (i %in% module_rows) {
    module_id <- paste0("md:", id)
    list(
      Input_ID = id,
      Type = "Module",
      Name = get_kegg_name(module_id),
      Modules = id,
      Pathways = paste(get_kegg_links("pathway", module_id), collapse = "; ")
    )
  } else if (i %in% pathway_rows) {
    pathway_id <- paste0("path:", id)
    list(
      Input_ID = id,
      Type = "Pathway",
      Name = get_kegg_name(pathway_id),
      Modules = NA,
      Pathways = id
    )
  } else {
    list(
      Input_ID = id,
      Type = "Unknown",
      Name = "unknown",
      Modules = NA,
      Pathways = NA
    )
  }
})

# Convert to data frame and save
kegg_results <- do.call(rbind, lapply(results, as.data.frame, stringsAsFactors = FALSE))
write.csv(kegg_results, "kegg_hierarchy_output.csv", row.names = FALSE)
